Enhanced execution guard

MishaDemianenko opened the pull request:
Enhance execution guard to be system wide.
Allow transactions and statement to have timeouts that can be global or instance specific.
Expose interface for external validation check of statements and transactions.
Remove custom server-based filters and guard manipulations.

changelog: Enhance execution guard to be applicable for all transactions
- Added `dbms.transaction.timeout` configuration option
- Allow starting transaction with custom timeout
- Deprecated `unsupported.dbms.executiontime_limit.time` configuration option
- Deprecated `unsupported.dbms.executiontime_limit.enabled` configuration option
- Renamed `dbms.transaction_timeout` configuration option to `dbms.rest.transaction.idle_timeout`

MishaDemianenko committed:
Port of 3.0 execution guard
Enhance execution guard to be system wide.
Allow transactions and statement to have timeouts that can be global or instance specific.
Expose interface for external validation check of statements and transactions.
Remove custom server-based filters and guard manipulations.

MishaDemianenko committed:
Allow to start transaction with custom timeout from GraphDatabaseFacade.
Rename property for timeout of idle transactions in the REST endpoint from 'dbms.transaction_timeout' to 'dbms.rest.transaction.idle_timeout'.

MishaDemianenko committed:
Remove `unsupported.dbms.executiontime_limit.enabled` system property.
Use positive value of `dbms.transaction.timeout` as indication of configured guard.

MishaDemianenko committed:
Always add guarding operations level to have possibility to terminate transaction with custom timeouts.
Allow execution of queries with custom timeout using GraphDatabaseService API.

MishaDemianenko committed:
Add additional IT tests for new api methods.
Cleanup.

burqen commented:
Changelog update ping

MishaDemianenko committed:
Add migration for changed 'dbms.transaction_timeout' server property.

MishaDemianenko committed:
Made test run faster.

MishaDemianenko committed:
Restore and deprecate obsolete execution guard properties.
Add migrations to those properties for more smooth user experience.

tinwelint commented:
@MishaDemianenko two test failures, where one seems trivial

MishaDemianenko committed:
Add support for custom transaction timeout into transactional endpoint.
Allow propagation of custom timeout if specified in transactional service endpoint.

Extract header extraction from service into custom utility class.

MishaDemianenko committed:
Add tests for periodic commit queries termination.

MishaDemianenko committed:
Temporary disable driver test

MishaDemianenko commented:
@benbc can you please take a look as well as operability team representative

benbc commented:
@MishaDemianenko Sorry we haven't got to this yet. I'm afraid we won't be able to do it before Friday.

tinwelint commented:
@MishaDemianenko needs rebase

MishaDemianenko commented:
@tinwelint rebased

MishaDemianenko commented:
@benbc that's fine, as soon as you have time guys even if it will be merged before we can do post updates

tinwelint commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java:
```diff
@@ -238,6 +238,23 @@
+     * @param timeout transaction timeout
+     * @return a new transaction instance
+     */
+    Transaction beginTx( long timeout );
```
Wouldn't it be nice and user-friendly with `long timeout, TimeUnit unit` ?

MishaDemianenko replied:
sounds like a good idea since then users do not need to perform conversion themself and that will be more readable

tinwelint commented on community/graphdb-api/src/main/java/org/neo4j/graphdb/GraphDatabaseService.java:
```diff
@@ -250,6 +267,19 @@
+     * @return A {@link org.neo4j.graphdb.Result} that contains the result set.
+     * @throws QueryExecutionException If the Query contains errors
+     */
+    Result execute( String query, long timeout ) throws QueryExecutionException;
```
Perhaps also unit here

tinwelint commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -201,6 +201,13 @@
 
+    @Internal
+    @Deprecated
+    public static final Setting<Boolean> execution_guard_enabled = setting("unsupported.dbms.executiontime_limit.enabled", BOOLEAN, FALSE );
```
is it really _unsupported_ btw?

benbc replied:
Or maybe the migration does that well enough already. Up to you.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/api/KernelTransaction.java:
```diff
@@ -171,7 +171,13 @@ default void close() throws TransactionFailureException
+
+    /**
+     * Timeout for transaction.
+     * @return transaction timeout
```
Include unit in javadoc

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -332,21 +343,33 @@ public Transaction beginTx()
-            // FIXME: perhaps we should check that the new type and access mode are compatible with the current tx
-            return new PlaceboTransaction( spi::currentTransaction, spi::currentStatement );
-        }
+        return beginTransaction( () -> spi.beginTransaction( type, accessMode ) );
```
Are we sure this won't result in `new` object allocation at runtime? We should generally avoid "creating" lambdas on hot paths unless we're sure what hotspot does to them.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -332,21 +343,33 @@ public Transaction beginTx()
+    public InternalTransaction beginTransaction( KernelTransaction.Type type, AccessMode accessMode,
+            long timeout )
+    {
+        return beginTransaction( () -> spi.beginTransaction( type, accessMode, timeout ) );
```
Same about lambdas as above

tinwelint commented on community/server/src/main/java/org/neo4j/server/database/CypherExecutor.java:
```diff
@@ -74,8 +80,24 @@ public void stop() throws Throwable
+    private InternalTransaction getInternalTransaction( HttpServletRequest request )
+    {
+        long customTimeout = getTransactionTimeout( request, log );
+        return customTimeout > 0 ? beginCustomTransaction( customTimeout ) : beginDefaultTransaction();
```
Isn't this an unnecessary check (change overall)? timeout: 0 means no timeout, doesn't it?

EDIT: oh I think I see the difference. To specifically provide 0 from here means that there will be no timeout, i.e. overriding any globally configured timeout... that's the difference between calling the method w/ and w/o, isn't it?

MishaDemianenko replied:
correct

tinwelint commented on community/server/src/main/java/org/neo4j/server/rest/transactional/TransitionalTxManagementKernelTransaction.java:
```diff
@@ -108,6 +110,12 @@ void closeTransactionForPeriodicCommit()
+
+    private InternalTransaction startTransaction()
+    {
+        return customTransactionTimeout > 0 ? db.beginTransaction( type, mode, customTransactionTimeout ) :
```
Also here... we could simply always call the method _with_ timeout, can't we?

tinwelint commented on community/server/src/main/java/org/neo4j/server/web/HttpHeaderUtils.java:
```diff
@@ -49,4 +54,30 @@ public static MediaType mediaTypeWithCharsetUtf8(MediaType mediaType)
+                        MAX_EXECUTION_TIME_HEADER, headerValue), e );
+            }
+        }
+        return -1;
```
Can't we use the usual `0` which means _no timeout_ in the rest of the code?

tinwelint replied:
Ping about this one. By the looks of it, the usages of this value never checks for -1, only `> 0` so I think it makes sense to use 0, potentially also with something like a `UNSPECIFIED_TIMEOUT` constant.

benbc commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -201,6 +201,13 @@
 
+    @Internal
+    @Deprecated
+    public static final Setting<Boolean> execution_guard_enabled = setting("unsupported.dbms.executiontime_limit.enabled", BOOLEAN, FALSE );
```
Please can you note what users of the deprecated setting should use instead. Also in the changelogs.

tinwelint replied:
I think we normally let the migration warning/info communicate that

benbc commented:
How does the timeout map to `org.neo4j.kernel.api.exceptions.Status`?

tinwelint commented:
@benbc what's your view on the `GraphDatabaseService#beginTx(timeout)` vs `GraphDatabaseService#beginTx(timeout,unit)` I raised above?

benbc commented:
@tinwelint I strongly agree with you, but I didn't want to put my oar in on API design.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -1002,13 +1007,10 @@ private StatementOperationParts buildStatementOperations(
-            parts = parts.override( null, null, guardingOperations, guardingOperations, null, null, null, null,
-                    null, null, null );
-        }
+        GuardingStatementOperations guardingOperations = new GuardingStatementOperations(
```
This will cause the guard layer to always be there, causing more delegation calls for all users. I think we should only have this layer on if there's a timeout configured.

MishaDemianenko replied:
You can always now ask for custom timeout even if default one is turned off, that's why we need that layer to be there

tinwelint replied:
Then perhaps we should have a `dbms.guard.enabled` setting as well, since it's unnecessary to have all users that don't use the guard pay for this extra layer. Ping @benbc on this.

Alternatively build two statement operation layers, one w/ guard and one w/o and use the appropriate one for each occasion... this is something we may want to do anyway, for example for read-only transactions where there's a couple of layers we don't actually need. Currently `KernelTransactions` accepts a single `StatementOperationParts` which contitutes the kernel cake, it could be made to accept multiple, although starting with these two would be enough. I'm adamant about not wasting CPU cycles on our hot paths, even if ever so few.

MishaDemianenko replied:
updated

MishaDemianenko committed:
Allow users to specify time unit together with timeout.

MishaDemianenko committed:
Add unit into javadoc

MishaDemianenko committed:
Avoid lambdas in GraphDatabaseFacade.
Always start transaction with timeout (default or custom).
Remove method to start transaction without timeout from SPI.

MishaDemianenko committed:
Introduce and use constant for unspecified timeout, update HttpHeaderUtils to use unspecified timeout constant instead of -1.

MishaDemianenko committed:
Add additional description what properties should be used instead of deprecated.

MishaDemianenko committed:
Introduce new TransactionTimedOut status to indicate cases when transaction has not completed within the specified timeout.
Add new status to guard exception.

MishaDemianenko committed:
Dynamically select operation parts: guarded or non guarded chain based on transaction timeout value.

MishaDemianenko commented:
@benbc thx about noticing missing status mapping. New `TransactionTimedOut` status introduced and used in guard exception.

MishaDemianenko commented:
@tinwelint ping, should be ready now

tinwelint commented:
@MishaDemianenko needs rebase, will merge on green now

MishaDemianenko commented:
@tinwelint rebased

tinwelint commented:
@MishaDemianenko Hmm, needs rebase again

MishaDemianenko committed:
Fix post merge problems

MishaDemianenko commented:
@tinwelint rebased again :)

MishaDemianenko commented:
@tinwelint ready now!
